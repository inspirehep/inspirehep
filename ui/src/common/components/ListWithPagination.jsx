import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { List, Pagination } from 'antd';
import Immutable from 'immutable';
import classNames from 'classnames';

const GRID_CONFIG = {
  gutter: 16,
  xs: 1,
  sm: 2,
  lg: 3,
  xl: 4,
};

/**
 * Only displays given page items at once and pagination ui
 * does not paginate.
 * Use `ClientPaginatedList` for complete solution, or implement server one like: `CitationList`
 */
class ListWithPagination extends Component {
  static getPaginationRangeInfo(total, range) {
    return `${range[0]}-${range[1]} of ${total}`;
  }

  renderPagination() {
    const { pageSize, loading, total, page, onPageChange, grid } = this.props;
    return (
      <Pagination
        className={classNames({ 'ant-col-24': grid })}
        hideOnSinglePage
        current={page}
        onChange={onPageChange}
        total={total}
        pageSize={pageSize}
        loading={loading}
        showTotal={ListWithPagination.getPaginationRangeInfo}
      />
    );
  }

  // wrap `
  render() {
    const { renderItem, title, pageItems, page, pageSize, grid } = this.props;
    return (
      <List
        header={title}
        footer={this.renderPagination()}
        grid={grid ? GRID_CONFIG : undefined}
      >
        <div
          className={
            classNames({
              'ant-row-flex': grid,
            }) /* to workaround ant-design/issues/14407 */
          }
        >
          {pageItems.map((item, index) =>
            renderItem(item, (page - 1) * pageSize + index)
          )}
        </div>
      </List>
    );
  }
}

ListWithPagination.propTypes = {
  total: PropTypes.number.isRequired,
  pageItems: PropTypes.instanceOf(Immutable.List).isRequired,
  renderItem: PropTypes.func.isRequired,
  onPageChange: PropTypes.func.isRequired,
  pageSize: PropTypes.number.isRequired,
  page: PropTypes.number.isRequired,
  title: PropTypes.node,
  loading: PropTypes.bool,
  grid: PropTypes.bool,
};

ListWithPagination.defaultProps = {
  title: null,
  loading: false,
};

export default ListWithPagination;
