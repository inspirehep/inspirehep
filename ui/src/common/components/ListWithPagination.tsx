import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { List, Pagination } from 'antd';
import Immutable from 'immutable';
import classNames from 'classnames';

const GRID_CONFIG = {
  gutter: 16,
  xs: 1,
  sm: 2,
  lg: 3,
  xl: 4,
};

const PAGE_SIZE_OPTIONS = ['25', '50', '100', '250'];

/**
 * Only displays given page items at once and pagination ui
 * does not paginate.
 * Use `ClientPaginatedList` for complete solution, or implement server one like: `CitationList`
 */
class ListWithPagination extends Component {
  static getPaginationRangeInfo(total: any, range: any) {
    return `${range[0]}-${range[1]} of ${total}`;
  }

  constructor(props: any) {
    super(props);

    this.renderItem = this.renderItem.bind(this);
  }

  renderPagination() {
    const {
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'pageSize' does not exist on type 'Readon... Remove this comment to see the full error message
      pageSize,
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'loading' does not exist on type 'Readonl... Remove this comment to see the full error message
      loading,
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'total' does not exist on type 'Readonly<... Remove this comment to see the full error message
      total,
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'page' does not exist on type 'Readonly<{... Remove this comment to see the full error message
      page,
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'onPageChange' does not exist on type 'Re... Remove this comment to see the full error message
      onPageChange,
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'grid' does not exist on type 'Readonly<{... Remove this comment to see the full error message
      grid,
      // @ts-expect-error ts-migrate(2339) FIXME: Property 'onSizeChange' does not exist on type 'Re... Remove this comment to see the full error message
      onSizeChange,
    } = this.props;
    return (
      <Pagination
        className={classNames({ 'ant-col-24': grid })}
        hideOnSinglePage
        current={page}
        onChange={onPageChange}
        total={total}
        pageSize={pageSize}
        // @ts-expect-error ts-migrate(2769) FIXME: No overload matches this call.
        loading={loading}
        showTotal={ListWithPagination.getPaginationRangeInfo}
        onShowSizeChange={onSizeChange}
        showSizeChanger={onSizeChange != null}
        pageSizeOptions={PAGE_SIZE_OPTIONS}
        responsive
      />
    );
  }

  renderItem(item: any, index: any) {
    // @ts-expect-error ts-migrate(2339) FIXME: Property 'renderItem' does not exist on type 'Read... Remove this comment to see the full error message
    const { renderItem, pageSize, page } = this.props;
    const absoluteIndex = (page - 1) * pageSize + index;
    return renderItem(item, absoluteIndex);
  }

  render() {
    // @ts-expect-error ts-migrate(2339) FIXME: Property 'title' does not exist on type 'Readonly<... Remove this comment to see the full error message
    const { title, grid, pageItems } = this.props;
    return (
      <List
        header={title}
        footer={this.renderPagination()}
        grid={grid ? GRID_CONFIG : undefined}
        dataSource={pageItems}
        renderItem={this.renderItem}
        data-test-id="pagination-list"
      />
    );
  }
}

// @ts-expect-error ts-migrate(2339) FIXME: Property 'propTypes' does not exist on type 'typeo... Remove this comment to see the full error message
ListWithPagination.propTypes = {
  total: PropTypes.number.isRequired,
  // @ts-expect-error ts-migrate(2345) FIXME: Argument of type 'typeof List' is not assignable t... Remove this comment to see the full error message
  pageItems: PropTypes.instanceOf(Immutable.List).isRequired,
  renderItem: PropTypes.func.isRequired,
  onPageChange: PropTypes.func.isRequired,
  onSizeChange: PropTypes.func,
  pageSize: PropTypes.number,
  page: PropTypes.number,
  title: PropTypes.node,
  loading: PropTypes.bool,
  grid: PropTypes.bool,
};

// @ts-expect-error ts-migrate(2339) FIXME: Property 'defaultProps' does not exist on type 'ty... Remove this comment to see the full error message
ListWithPagination.defaultProps = {
  title: null,
  loading: false,
  page: 1,
  pageSize: 25,
};

export default ListWithPagination;
