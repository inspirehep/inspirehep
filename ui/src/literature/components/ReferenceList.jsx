import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { List, Map } from 'immutable';

import ClientPaginatedList from '../../common/components/ClientPaginatedList';
import ContentBox from '../../common/components/ContentBox';
import ReferenceItem from './ReferenceItem';
import ErrorAlertOrChildren from '../../common/components/ErrorAlertOrChildren';
import EmptyOrChildren from '../../common/components/EmptyOrChildren';

class ReferenceList extends Component {
  static renderReferenceItem(reference, index, page) {
    return (
      // reference data model doesn't have any identifier, thus we have hack for `key`
      <ReferenceItem
        key={reference.getIn(['titles', 0, 'title']) || String(index * page)}
        reference={reference}
      />
    );
  }

  renderList() {
    const { references } = this.props;
    return (
      references && (
        <ClientPaginatedList
          items={references}
          renderItem={ReferenceList.renderReferenceItem}
        />
      )
    );
  }

  render() {
    const { loading, error, references } = this.props;
    return (
      <ContentBox loading={loading}>
        <ErrorAlertOrChildren error={error}>
          <EmptyOrChildren data={references} description="0 References">
            {this.renderList()}
          </EmptyOrChildren>
        </ErrorAlertOrChildren>
      </ContentBox>
    );
  }
}

ReferenceList.propTypes = {
  references: PropTypes.instanceOf(List),
  error: PropTypes.instanceOf(Map),
  loading: PropTypes.bool,
};

ReferenceList.defaultProps = {
  references: null,
  error: null,
  loading: false,
};

export default ReferenceList;
