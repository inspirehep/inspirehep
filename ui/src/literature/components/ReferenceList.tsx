import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { List } from 'immutable';

import ContentBox from '../../common/components/ContentBox';
import ReferenceItem from './ReferenceItem';
import ErrorAlertOrChildren from '../../common/components/ErrorAlertOrChildren';
import EmptyOrChildren from '../../common/components/EmptyOrChildren';
import { ErrorPropType } from '../../common/propTypes';
import ListWithPagination from '../../common/components/ListWithPagination';

function ReferenceList({
  total,
  references,
  error,
  loading,
  onQueryChange,
  query,
}) {
  const renderReferenceItem = useCallback(
    (reference, index) => (
      // reference data model doesn't have any identifier, thus we have hack for `key`
      // FIXME: find an proper key for reference item as index might cause bugs
      <ReferenceItem
        key={reference.getIn(['titles', 0, 'title']) || String(index)}
        reference={reference}
      />
    ),
    []
  );

  const onPageChange = useCallback(page => onQueryChange({ page }), [
    onQueryChange,
  ]);

  const onSizeChange = useCallback(
    (page, size) => onQueryChange({ size, page: '1' }),
    [onQueryChange]
  );

  const renderList = useCallback(
    () =>
      total > 0 && (
        <ListWithPagination
          renderItem={renderReferenceItem}
          pageItems={references}
          onPageChange={onPageChange}
          onSizeChange={onSizeChange}
          total={total}
          page={query.page}
          pageSize={query.size}
        />
      ),
    [
      query.page,
      query.size,
      total,
      references,
      renderReferenceItem,
      onPageChange,
      onSizeChange,
    ]
  );

  return (
    <ContentBox loading={loading}>
      <ErrorAlertOrChildren error={error}>
        <EmptyOrChildren data={references} title="0 References">
          {renderList()}
        </EmptyOrChildren>
      </ErrorAlertOrChildren>
    </ContentBox>
  );
}
ReferenceList.propTypes = {
  total: PropTypes.number.isRequired,
  references: PropTypes.instanceOf(List).isRequired,
  error: ErrorPropType,
  loading: PropTypes.bool.isRequired,
  onQueryChange: PropTypes.func.isRequired,
  query: PropTypes.objectOf(PropTypes.any).isRequired,
};

export default ReferenceList;
