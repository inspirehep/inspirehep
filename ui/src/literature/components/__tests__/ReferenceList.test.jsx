import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import ReferenceList from '../ReferenceList';
import ListWithPagination from '../../../common/components/ListWithPagination';

describe('ReferenceList', () => {
  it('renders with references', () => {
    const references = fromJS([
      {
        titles: [{ title: 'Reference 1' }],
      },
      {
        titles: [{ title: 'Reference 2' }],
      },
    ]);
    const wrapper = shallow(
      <ReferenceList
        loading={false}
        error={null}
        references={references}
        total={1}
        onPageChange={() => {}}
        onSizeChange={() => {}}
        query={{ size: 25, page: 1 }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders items with (page * index) key if title is absent', () => {
    const references = fromJS([
      {
        publication_info: [{ journal_title: 'Journal 1' }],
      },
      {
        authors: [{ full_name: 'Author 2' }],
      },
    ]);
    const wrapper = shallow(
      <ReferenceList
        loading={false}
        error={null}
        references={references}
        total={1}
        onPageChange={() => {}}
        onSizeChange={() => {}}
        query={{ size: 25, page: 1 }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('calls onQueryChange and sets the correct page', () => {
    const onPageChange = jest.fn();
    const wrapper = shallow(
      <ReferenceList
        loading={false}
        error={null}
        references={fromJS([{ titles: [{ title: 'Reference 1' }] }])}
        total={50}
        onPageChange={onPageChange}
        query={{ size: 25, page: 1 }}
      />
    );
    const page = 2;
    const size = 25;
    const onListPageChange = wrapper
      .find(ListWithPagination)
      .prop('onPageChange');
    onListPageChange(page, size);
    expect(onPageChange).toHaveBeenCalledWith(page, size);
  });

  it('does not render the list if total 0', () => {
    const wrapper = shallow(
      <ReferenceList
        loading={false}
        error={null}
        references={fromJS([{ titles: [{ title: 'Reference 1' }] }])}
        total={0}
        onPageChange={() => {}}
        query={{ size: 25, page: 1 }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with error', () => {
    const wrapper = shallow(
      <ReferenceList
        loading={false}
        error={fromJS({ message: 'error' })}
        references={fromJS([])}
        total={0}
        onPageChange={() => {}}
        query={{ size: 25, page: 1 }}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});

it('calls onQueryChange and sets display to 50 references/page', () => {
  const onSizeChange = jest.fn();
  const wrapper = shallow(
    <ReferenceList
      loading={false}
      error={null}
      references={fromJS([{ titles: [{ title: 'Reference 1' }] }])}
      total={50}
      onPageChange={() => {}}
      onSizeChange={onSizeChange}
      query={{ size: 50 }}
    />
  );
  const page = '1';
  const size = 50;
  const onListPageSizeChange = wrapper
    .find(ListWithPagination)
    .prop('onSizeChange');
  onListPageSizeChange(page, size);
  expect(onSizeChange).toHaveBeenCalledWith(page, size);
});
