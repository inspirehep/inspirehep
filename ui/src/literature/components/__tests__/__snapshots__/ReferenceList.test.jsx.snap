// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReferenceList does not render without references 1`] = `
<ContentBox
  className=""
  leftActions={null}
  loading={false}
  rightActions={null}
  subTitle={null}
  title={null}
>
  <ErrorAlertOrChildren
    error={null}
  />
</ContentBox>
`;

exports[`ReferenceList renders as loading if set 1`] = `
<ContentBox
  className=""
  leftActions={null}
  loading={true}
  rightActions={null}
  subTitle={null}
  title={null}
>
  <ErrorAlertOrChildren
    error={null}
  >
    <ClientPaginatedList
      items={Immutable.List []}
      loading={false}
      pageSize={25}
      renderItem={[Function]}
      title={null}
    />
  </ErrorAlertOrChildren>
</ContentBox>
`;

exports[`ReferenceList renders items with (page * index) key if title is absent 1`] = `
<div
  className="__ContentBox__ h-100 "
>
  <Card
    className="h-100"
    loading={false}
    title={null}
  >
    <div
      className="pa2"
    >
      <div>
        <ErrorAlertOrChildren
          error={null}
        >
          <ClientPaginatedList
            items={
              Immutable.List [
                Immutable.Map {
                  "publication_info": Immutable.List [
                    Immutable.Map {
                      "journal_title": "Journal 1",
                    },
                  ],
                },
                Immutable.Map {
                  "authors": Immutable.List [
                    Immutable.Map {
                      "full_name": "Author 2",
                    },
                  ],
                },
              ]
            }
            loading={false}
            pageSize={25}
            renderItem={[Function]}
            title={null}
          />
        </ErrorAlertOrChildren>
      </div>
    </div>
    <Row
      className="actions ph2"
      gutter={0}
      justify="space-between"
      type="flex"
    >
      <Col />
      <Col />
    </Row>
  </Card>
</div>
`;

exports[`ReferenceList renders with error 1`] = `
<ContentBox
  className=""
  leftActions={null}
  loading={false}
  rightActions={null}
  subTitle={null}
  title={null}
>
  <ErrorAlertOrChildren
    error={
      Immutable.Map {
        "message": "Error",
      }
    }
  >
    <ClientPaginatedList
      items={Immutable.List []}
      loading={false}
      pageSize={25}
      renderItem={[Function]}
      title={null}
    />
  </ErrorAlertOrChildren>
</ContentBox>
`;

exports[`ReferenceList renders with references 1`] = `
<div
  className="__ContentBox__ h-100 "
>
  <Card
    className="h-100"
    loading={false}
    title={null}
  >
    <div
      className="pa2"
    >
      <div>
        <ErrorAlertOrChildren
          error={null}
        >
          <ClientPaginatedList
            items={
              Immutable.List [
                Immutable.Map {
                  "titles": Immutable.List [
                    Immutable.Map {
                      "title": "Reference 1",
                    },
                  ],
                },
                Immutable.Map {
                  "titles": Immutable.List [
                    Immutable.Map {
                      "title": "Reference 2",
                    },
                  ],
                },
              ]
            }
            loading={false}
            pageSize={25}
            renderItem={[Function]}
            title={null}
          />
        </ErrorAlertOrChildren>
      </div>
    </div>
    <Row
      className="actions ph2"
      gutter={0}
      justify="space-between"
      type="flex"
    >
      <Col />
      <Col />
    </Row>
  </Card>
</div>
`;
